File: apollo/src/main/java/io/muun/apollo/data/net/base/interceptor/AuthHeaderInterceptor.java
package io.muun.apollo.data.net.base.interceptor;

import io.muun.apollo.data.net.base.BaseInterceptor;
import io.muun.apollo.data.preferences.AuthRepository;
import io.muun.common.net.HeaderUtils;

import okhttp3.Request;
import okhttp3.Response;

import javax.inject.Inject;

public class AuthHeaderInterceptor extends BaseInterceptor {

    private final AuthRepository authRepository;

    @Inject
    public AuthHeaderInterceptor(AuthRepository authRepository) {
        this.authRepository = authRepository;
    }

    @Override
    protected Request processRequest(Request request) {
        // Attach the request header if a token is available:

        final String serverJwt = authRepository.getServerJwt().orElse(null);

        if (serverJwt != null) {
            return request.newBuilder()
                    .addHeader(HeaderUtils.AUTHORIZATION, "Bearer " + serverJwt)
                    .build();

        } else {
            return request;
        }
    }

    @Override
    protected Response processResponse(Response response) {
        // Save the token in the response header if one is found:
        final String authHeaderValue = response.header(HeaderUtils.AUTHORIZATION);

        HeaderUtils.getBearerTokenFromHeader(authHeaderValue)
                .ifPresent(authRepository::storeServerJwt);

        // We need a reliable way (across all envs: local, CI, prd, etc...) to identify the logout
        // requests. We could inject HoustonConfig and build the entire URL (minus port)
        // or... we can do this :)
        final String url = response.request().url().url().toString();
        final boolean isLogout = url.endsWith("sessions/logout");

        if (!isLogout) {
            final String sessionStatusHeaderValue = response.header(HeaderUtils.SESSION_STATUS);

            HeaderUtils.getSessionStatusFromHeader(sessionStatusHeaderValue)
                    .ifPresent(authRepository::storeSessionStatus);
        }

        return response;
    }
}

File: android/apollo/src/main/java/io/muun/apollo/data/net/base/interceptor/AuthHeaderInterceptor.kt
package io.muun.apollo.data.net.base.interceptor

import io.muun.apollo.data.net.base.BaseInterceptor
import io.muun.apollo.data.preferences.AuthRepository
import io.muun.common.model.SessionStatus
import io.muun.common.net.HeaderUtils
import okhttp3.Request
import okhttp3.Response
import javax.inject.Inject

class AuthHeaderInterceptor @Inject constructor(
    private val authRepository: AuthRepository,
) : BaseInterceptor() {

    override fun processRequest(request: Request): Request {
        // Attach the request header if a token is available:
        val serverJwt = authRepository.serverJwt.orElse(null)
        return if (serverJwt != null) {
            request.newBuilder()
                .addHeader(HeaderUtils.AUTHORIZATION, "Bearer $serverJwt")
                .build()

        } else {
            request
        }
    }

    override fun processResponse(response: Response): Response {
        // Save the token in the response header if one is found
        val authHeaderValue = response.header(HeaderUtils.AUTHORIZATION)
        HeaderUtils.getBearerTokenFromHeader(authHeaderValue)
            .ifPresent { serverJwt: String -> authRepository.storeServerJwt(serverJwt) }

        // We need a reliable way (across all envs: local, CI, prd, etc...) to identify the logout
        // requests. We could inject HoustonConfig and build the entire URL (minus port)
        // or... we can do this :)
        val url = response.request().url().url().toString()
        val isLogout = url.endsWith("sessions/logout")

        if (!isLogout) {
            val sessionStatusHeaderValue = response.header(HeaderUtils.SESSION_STATUS)
            HeaderUtils.getSessionStatusFromHeader(sessionStatusHeaderValue)
                .ifPresent { sessionStatus: SessionStatus ->
                    authRepository.storeSessionStatus(sessionStatus)
                }
        }
        return response
    }
}